<a href="https://www.kaggle.com/code/rickyricardo7467/data-cleaning?scriptVersionId=199114032" target="_blank"><img align="left" alt="Kaggle" title="Open in Kaggle" src="https://kaggle.com/static/images/open-in-kaggle.svg"></a>

# Import Library's
We will use pandas to process and read our banking dataset. 

In [122]:

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))


/kaggle/input/churn-modeling-4/Churn_Modeling_3-2.csv
/kaggle/input/churndata2/Churn_Modeling_2.csv
/kaggle/input/banking-clean/Churn_Modelling.csv


# Examining the Data
I'll use python and use the commands data.head, data.describe and data.info to basic statistics in the first few rows of the dataset. This will give me a good idea of the types of data and Column names I am going to be working with. 

In [123]:
data = pd.read_csv('/kaggle/input/banking-clean/Churn_Modelling.csv')
data.head()
data.describe()
data.info ()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10002 entries, 0 to 10001
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CustomerId       10002 non-null  int64  
 1   Surname          10002 non-null  object 
 2   CreditScore      10002 non-null  int64  
 3   Geography        10001 non-null  object 
 4   Gender           10002 non-null  object 
 5   Age              10001 non-null  float64
 6   Tenure           10002 non-null  int64  
 7   Balance          10002 non-null  float64
 8   NumOfProducts    10002 non-null  int64  
 9   HasCrCard        10001 non-null  float64
 10  IsActiveMember   10001 non-null  float64
 11  EstimatedSalary  10002 non-null  float64
 12  Exited           10002 non-null  int64  
dtypes: float64(5), int64(5), object(3)
memory usage: 1016.0+ KB


# Identify Missing Values
To check for missing values, I will use the data.isnull command to show the total missing values of each column. Ill also add the data.info to show what the datatype is that has missing values. 

In [124]:
print(data.isnull().sum())
data.info()

CustomerId         0
Surname            0
CreditScore        0
Geography          1
Gender             0
Age                1
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          1
IsActiveMember     1
EstimatedSalary    0
Exited             0
dtype: int64
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10002 entries, 0 to 10001
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CustomerId       10002 non-null  int64  
 1   Surname          10002 non-null  object 
 2   CreditScore      10002 non-null  int64  
 3   Geography        10001 non-null  object 
 4   Gender           10002 non-null  object 
 5   Age              10001 non-null  float64
 6   Tenure           10002 non-null  int64  
 7   Balance          10002 non-null  float64
 8   NumOfProducts    10002 non-null  int64  
 9   HasCrCard        10001 non-null  float64
 10  IsActiveMember   10001 non-null  float64
 11  Es

# Missing Values by row
I can also help narrow down where the values are missing by row using the following python command. 

In [125]:
print(data[data.isnull().any(axis=1)])

   CustomerId   Surname  CreditScore Geography  Gender   Age  Tenure  \
4    15737888  Mitchell          850     Spain  Female  43.0       2   
6    15592531  Bartlett          822       NaN    Male  50.0       7   
8    15792365        He          501    France    Male  44.0       4   
9    15592389        H?          684    France    Male   NaN       2   

     Balance  NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  \
4  125510.82              1        NaN             1.0         79084.10   
6       0.00              2        1.0             1.0         10062.80   
8  142051.07              2        0.0             NaN         74940.50   
9  134603.88              1        1.0             1.0         71725.73   

   Exited  
4       0  
6       0  
8       0  
9       0  


As we can see we have a strange Surname for Customer ID #15592389, we have a few missing values under "Country," "Age," "Has Credit Card" and "Is Active Memeber."

# Addressing Missing Values
Here we can go in and address each value. 
1. Although it was not techinically "missing", we will start with replacing the "H?" surname with "Henry."
2. Next we will move onto the missing values by Country. 
    * I will replace the single missing value in with the most commonly used country. 
3. Next we have a missing value in the age column we will replace with the average age of the entire column. 
4. To address the missing value in "HasCrCard" and "IsActiveMember" we will fill it with the value "Yes"(1) or "No"(0).
5. After all of our Missing values have been addressed, we need to verify the changes prior to moving to the next step in the Data cleaning process. 

In [126]:
#Replacing the "H?" with "Henry"
data['Surname'].replace('H?', 'Henry', inplace=True)


/tmp/ipykernel_30/2563861737.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Surname'].replace('H?', 'Henry', inplace=True)


In [127]:
#Replacing the missing values in Geography with the most commonly used Country.
data['Geography'].fillna(data['Geography'].mode()[0], inplace=True)

/tmp/ipykernel_30/833773543.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Geography'].fillna(data['Geography'].mode()[0], inplace=True)


In [128]:
#Replacing the missing value in "Age" with the average age of the entire column. 
data['Age'].fillna(data['Age'].mean(), inplace=True)


/tmp/ipykernel_30/4001107021.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(data['Age'].mean(), inplace=True)


In [129]:
# Replace missing values with 1 (Yes)
data['HasCrCard'].fillna(1, inplace=True)

/tmp/ipykernel_30/256288808.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['HasCrCard'].fillna(1, inplace=True)


In [130]:
# Replace missing values with 0 (No)
data['HasCrCard'].fillna(0, inplace=True)

/tmp/ipykernel_30/4041130040.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['HasCrCard'].fillna(0, inplace=True)


In [131]:
# Replace missing values with 1 (Yes)
data['IsActiveMember'].fillna(1, inplace=True)  

/tmp/ipykernel_30/152668195.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['IsActiveMember'].fillna(1, inplace=True)


In [132]:
# Replace missing values with 0 (No)
data['IsActiveMember'].fillna(0, inplace=True)  

/tmp/ipykernel_30/1279710514.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['IsActiveMember'].fillna(0, inplace=True)


In [133]:
#To verify our missing values have been handled we are going to run the print command to ensure they have been handled. 
print(data.isnull().sum())

CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64


In [134]:
# Display DataFrame after filling missing values
print("\nDataFrame After Filling Missing Values:\n", data)


DataFrame After Filling Missing Values:
        CustomerId    Surname  CreditScore Geography  Gender   Age  Tenure  \
0        15634602   Hargrave          619    France  Female  42.0       2   
1        15647311       Hill          608     Spain  Female  41.0       1   
2        15619304       Onio          502    France  Female  42.0       8   
3        15701354       Boni          699    France  Female  39.0       1   
4        15737888   Mitchell          850     Spain  Female  43.0       2   
...           ...        ...          ...       ...     ...   ...     ...   
9997     15584532        Liu          709    France  Female  36.0       7   
9998     15682355  Sabbatini          772   Germany    Male  42.0       3   
9999     15682355  Sabbatini          772   Germany    Male  42.0       3   
10000    15628319     Walker          792    France  Female  28.0       4   
10001    15628319     Walker          792    France  Female  28.0       4   

         Balance  NumOfProducts  

In [135]:
# Save the DataFrame as a CSV file
data.to_csv('Churn_Modeling_2.csv', index=False)
print("Data has been cleaned and saved successfully!")

Data has been cleaned and saved successfully!


# Identifying and Handling Outliers: 

I will start with addressing numerical outliers on the following columns:
CreditScore:
Outliers might indicate customers with unusually low or high creditworthiness.

Age:
Outliers could indicate very young or very old customers. I will take into consideration which age groups I would want to include. 

Tenure:
Outliers may not always indicate an issue. Tenure should be analyzed based on business context.

Balance:
High outliers could be wealthy clients, but extremely low negative balances might indicate errors.

NumOfProducts:
If some customers have unusually high numbers of products, it might be useful to keep those data points.

EstimatedSalary:
High salary outliers could represent high-net-worth individuals. 



# Identify Outliers Using the Interquartile Range (IQR) Method

I will be using the IQR method to identify my numerical outliers. The IQR identifies outliers based on the range between the first quartile (Q1) and third quartile (Q3).

Formula for IQR:

	•	IQR = Q3 - Q1
	•	Lower Bound = Q1 - 1.5 * IQR
	•	Upper Bound = Q3 + 1.5 * IQR

Any data points outside these bounds are considered outliers.

In [136]:
import pandas as pd
data = pd.read_csv('/kaggle/input/churndata2/Churn_Modeling_2.csv')

# List of numerical columns to check for outliers
numeric_columns = ['CreditScore', 'Age', 'Tenure', 'Balance', 'NumOfProducts', 'EstimatedSalary']

# Detect outliers using IQR method for each numerical column
for column in numeric_columns:
    Q1 = data[column].quantile(0.25)  # First quartile (25th percentile)
    Q3 = data[column].quantile(0.75)  # Third quartile (75th percentile)
    IQR = Q3 - Q1

    # Define lower and upper bounds
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # Identify outliers
    outliers = data[(data[column] < lower_bound) | (data[column] > upper_bound)]
    print(f"Column: {column}")
    print(f"Number of outliers: {outliers.shape[0]}")
    print(outliers[[column]])  # Display outlier values
    print("----" * 10)

Column: CreditScore
Number of outliers: 15
      CreditScore
7             376
942           376
1193          363
1405          359
1631          350
1838          350
1962          358
2473          351
2579          365
8154          367
8723          350
8762          350
9210          382
9356          373
9624          350
----------------------------------------
Column: Age
Number of outliers: 359
       Age
58    66.0
85    75.0
104   65.0
158   73.0
181   65.0
...    ...
9753  68.0
9765  64.0
9832  64.0
9894  77.0
9936  77.0

[359 rows x 1 columns]
----------------------------------------
Column: Tenure
Number of outliers: 0
Empty DataFrame
Columns: [Tenure]
Index: []
----------------------------------------
Column: Balance
Number of outliers: 0
Empty DataFrame
Columns: [Balance]
Index: []
----------------------------------------
Column: NumOfProducts
Number of outliers: 60
      NumOfProducts
7                 4
70                4
1254              4
1469              4
1488

For an additional examination of the outliers, I wanted to see the top highest and lowest outliers for each column. 

In [137]:
def view_top_outliers(df):
    # Identify numerical columns
    numerical_columns = df.select_dtypes(include=['float64', 'int64']).columns

    # Loop through each numerical column
    for column in numerical_columns:
        # Calculate Q1, Q3, and IQR
        Q1 = df[column].quantile(0.25)
        Q3 = df[column].quantile(0.75)
        IQR = Q3 - Q1

        # Define bounds for outliers
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR

        # Filter to identify outliers
        outliers = df[(df[column] < lower_bound) | (df[column] > upper_bound)]

        # Display top 10 lowest and highest outliers for each column
        if not outliers.empty:
            print(f"Top 10 Lowest Outliers for {column}:")
            print(outliers.nsmallest(10, column)[[column]], "\n")

            print(f"Top 10 Highest Outliers for {column}:")
            print(outliers.nlargest(10, column)[[column]], "\n")
            print("-" * 40)

# Step 2: Call the function on the DataFrame
view_top_outliers(data)

Top 10 Lowest Outliers for CreditScore:
      CreditScore
1631          350
1838          350
8723          350
8762          350
9624          350
2473          351
1962          358
1405          359
1193          363
2579          365 

Top 10 Highest Outliers for CreditScore:
      CreditScore
9210          382
7             376
942           376
9356          373
8154          367
2579          365
1193          363
1405          359
1962          358
2473          351 

----------------------------------------
Top 10 Lowest Outliers for Age:
       Age
567   63.0
769   63.0
963   63.0
1235  63.0
1285  63.0
2053  63.0
2078  63.0
2459  63.0
2615  63.0
2926  63.0 

Top 10 Highest Outliers for Age:
       Age
6443  92.0
6759  92.0
2458  88.0
3033  85.0
3387  84.0
3531  84.0
9309  83.0
766   82.0
4931  81.0
7526  81.0 

----------------------------------------
Top 10 Lowest Outliers for NumOfProducts:
      NumOfProducts
7                 4
70                4
1254              4
1469

# Removing Numerical Outliers
I have decided to remove unrealistically low credit scores. 

In [138]:
data_cleaned = data[(data['CreditScore'] >= lower_bound) & (data['CreditScore'] <= upper_bound)]

# Step 3: View the cleaned DataFrame
print("Original DataFrame:\n", data)
print("\nDataFrame after removing CreditScore outliers:\n", data_cleaned)

# Save the DataFrame as a CSV file
data.to_csv('Churn_Modeling_3.csv', index=False)
print("Data has been cleaned and saved successfully!")

Original DataFrame:
        CustomerId    Surname  CreditScore Geography  Gender   Age  Tenure  \
0        15634602   Hargrave          619    France  Female  42.0       2   
1        15647311       Hill          608     Spain  Female  41.0       1   
2        15619304       Onio          502    France  Female  42.0       8   
3        15701354       Boni          699    France  Female  39.0       1   
4        15737888   Mitchell          850     Spain  Female  43.0       2   
...           ...        ...          ...       ...     ...   ...     ...   
9997     15584532        Liu          709    France  Female  36.0       7   
9998     15682355  Sabbatini          772   Germany    Male  42.0       3   
9999     15682355  Sabbatini          772   Germany    Male  42.0       3   
10000    15628319     Walker          792    France  Female  28.0       4   
10001    15628319     Walker          792    France  Female  28.0       4   

         Balance  NumOfProducts  HasCrCard  IsActiveMe

In [139]:
#Again lets double check and make sure our values have been handled properly after the save. 
import pandas as pd
missing_values = data.isnull().sum()
print("Missing values in each column:\n", missing_values)

Missing values in each column:
 CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender             0
Age                0
Tenure             0
Balance            0
NumOfProducts      0
HasCrCard          0
IsActiveMember     0
EstimatedSalary    0
Exited             0
dtype: int64


In [140]:
#Now I am going to use IQR to remove those credit score outliers. 
import pandas as pd

# Load your dataset (replace 'your_dataset.csv' with your actual file name)
data = pd.read_csv('/kaggle/input/churn-modeling-4/Churn_Modeling_3-2.csv')

# Calculate the IQR for the CreditScore column
Q1 = data['CreditScore'].quantile(0.25)  # 25th percentile
Q3 = data['CreditScore'].quantile(0.75)  # 75th percentile
IQR = Q3 - Q1  # Interquartile Range

# Define lower and upper bounds for outliers
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

# Remove rows with CreditScore values outside the bounds
data_no_outliers = data[(data['CreditScore'] >= lower_bound) & (data['CreditScore'] <= upper_bound)]

# Print the number of removed outliers
print(f"Number of outliers removed: {len(data) - len(data_no_outliers)}")

Number of outliers removed: 15


In [141]:
# Display the cleaned DataFrame
print("DataFrame after removing outliers from CreditScore column:")
print(data_no_outliers.head())  # Display the first 5 rows of the cleaned DataFrame

DataFrame after removing outliers from CreditScore column:
   CustomerId   Surname  CreditScore Geography  Gender  Age  Tenure  \
0    15634602  Hargrave          619    France  Female   42       2   
1    15647311      Hill          608     Spain  Female   41       1   
2    15619304      Onio          502    France  Female   42       8   
3    15701354      Boni          699    France  Female   39       1   
4    15737888  Mitchell          850     Spain  Female   43       2   

     Balance  NumOfProducts  HasCrCard  IsActiveMember  EstimatedSalary  \
0       0.00              1          1               1        101348.88   
1   83807.86              1          0               1        112542.58   
2  159660.80              3          1               0        113931.57   
3       0.00              2          0               0         93826.63   
4  125510.82              1          1               1         79084.10   

   Exited  
0       1  
1       0  
2       1  
3       0  
4  

In [142]:
# Save the cleaned DataFrame to a new CSV file
data.to_csv('Churn_Modeling_3.csv', index=False)

# Print confirmation
print("Cleaned data has been saved to 'cleaned_data_no_outliers.csv'.")

Cleaned data has been saved to 'cleaned_data_no_outliers.csv'.


# Checking for Data Imbalance
I want to check the distribution on Geography and Gender and check how frequently Exited, "IsActiveMember," "HasCrCard," and 

In [143]:
# Check frequency distribution for categorical columns
print(data['Geography'].value_counts())
print(data['Gender'].value_counts())
print("Class distribution in target variable:\n", data_no_outliers['Exited'].value_counts())
print("Class distribution in target variable:\n", data_no_outliers['IsActiveMember'].value_counts())
print("Class distribution in target variable:\n", data_no_outliers['HasCrCard'].value_counts())


Geography
France     5015
Germany    2510
Spain      2477
Name: count, dtype: int64
Gender
Male      5458
Female    4544
Name: count, dtype: int64
Class distribution in target variable:
 Exited
0    7964
1    2023
Name: count, dtype: int64
Class distribution in target variable:
 IsActiveMember
1    5147
0    4840
Name: count, dtype: int64
Class distribution in target variable:
 HasCrCard
1    7044
0    2943
Name: count, dtype: int64


# Final Verification
Before I can move onto using my cleaned dataset, I am going to verify our outliers and missing values have been handled

In [144]:
print("Final cleaned data summary:")
print(data_no_outliers.info())
print("\nMissing values:\n", data_no_outliers.isnull().sum())

Final cleaned data summary:
<class 'pandas.core.frame.DataFrame'>
Index: 9987 entries, 0 to 10001
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   CustomerId       9987 non-null   int64  
 1   Surname          9987 non-null   object 
 2   CreditScore      9987 non-null   int64  
 3   Geography        9987 non-null   object 
 4   Gender           9987 non-null   object 
 5   Age              9987 non-null   int64  
 6   Tenure           9987 non-null   int64  
 7   Balance          9987 non-null   float64
 8   NumOfProducts    9987 non-null   int64  
 9   HasCrCard        9987 non-null   int64  
 10  IsActiveMember   9987 non-null   int64  
 11  EstimatedSalary  9987 non-null   float64
 12  Exited           9987 non-null   int64  
dtypes: float64(2), int64(8), object(3)
memory usage: 1.1+ MB
None

Missing values:
 CustomerId         0
Surname            0
CreditScore        0
Geography          0
Gender    

In [145]:
#One last save before exporting my dataset. 
#Hope this helped and if you have any suggestions feel free to let me know. 
# Save the cleaned DataFrame to a new CSV file
data.to_csv('Final_Cleaned_Banking_Dataset.csv', index=False)

# Print confirmation
print("Cleaned data has been saved to 'cleaned_data_no_outliers.csv'.")

Cleaned data has been saved to 'cleaned_data_no_outliers.csv'.
